### PR TITLE
fix(executors): fix gpu supports

### LINKED ISSUE #352: Running transformer model on GPU
I would like to be able to use my GPU to generate sentence embeddings. But as far as I can see from [transformer.py](https://github.com/jina-ai/jina/blob/9157e1a9429ccd04c7e5123273d40b136f69d977/jina/executors/encoders/nlp/transformer.py) code there's currently no way to make it happen. I would also like to have some more documentation on `on_gpu` meta.

**Environment**
```
jina                          0.1.2
jina-proto                    0.0.20
jina-vcs-tag                  (unset)
libzmq                        4.3.2
pyzmq                         1.18.3
protobuf                      3.11.3
proto-backend                 cpp
grpcio                        1.28.1
ruamel.yaml                   0.16.10
python                        3.8.2
platform                      Linux
platform-release              5.4.0-28-generic
platform-version              #32-Ubuntu SMP Wed Apr 22 17:40:10 UTC 2020
architecture                  x86_64
processor                     x86_64
jina-resources                /home/v/jina/examples/venv/lib/python3.8/site-packages/jina/resources
JINA_ARRAY_QUANT              (unset)
JINA_CONTRIB_MODULE           (unset)
JINA_CONTRIB_MODULE_IS_LOADING(unset)
JINA_CONTROL_PORT             (unset)
JINA_DEFAULT_HOST             (unset)
JINA_EXECUTOR_WORKDIR         (unset)
JINA_FULL_CLI                 (unset)
JINA_IPC_SOCK_TMP             (unset)
JINA_LOG_FILE                 (unset)
JINA_LOG_LONG                 (unset)
JINA_LOG_NO_COLOR             (unset)
JINA_LOG_PROFILING            (unset)
JINA_LOG_SSE                  (unset)
JINA_LOG_VERBOSITY            (unset)
JINA_PROFILING                (unset)
JINA_SOCKET_HWM               (unset)
JINA_STACK_CONFIG             (unset)
JINA_TEST_CONTAINER           (unset)
JINA_TEST_PRETRAINED          (unset)
JINA_VCS_VERSION              (unset)
JINA_VERSION                  (unset)
JINA_WARN_UNNAMED             (unset)
```

### PR DESCRIPTION
**Pull request by nan-wang**
resolves #352 

- fix the support for GPUs
- refactoring executors codes

### CODE REVIEW

File: jina/executors/__init__.py [hunk 1]
```diff
@@ -536,3 +556,82 @@ def __call__(self, req_type, *args, **kwargs):
...
+    """
+    :class:`BaseFrameworkExecutor` is the base class for the executors using other frameworks internally, including
+        `tensorflow`, `pytorch`, `onnx`, and, `paddlepaddle`.
+
+    ..notes:
+        The derived classes must implement `build_model()` and `set_device()` methods.
+    """
+    def post_init(self):
+        super().post_init()
+        self.pre_set_device()
```
Comment (hanxiao) [inline jina/executors/__init__.py:571]: if `pre_set_device` and `post_set_device` is an "either-or", then use a meta tag to branch the logics.
e.g. add a meta tag `gpu_placement`: `before`/`after`

File: jina/executors/encoders/image/tfkeras.py [hunk 2]
```diff
@@ -42,14 +43,13 @@ def __init__(self, model_name: str = 'MobileNetV2', img_shape: int = 96,
         self.img_shape = img_shape
         self.channel_axis = channel_axis
 
-    def post_init(self):
+    def build_model(self):
```
Comment (hanxiao) [inline jina/executors/encoders/image/tfkeras.py:46]: please don't change the pattern of `post_init`.

File: jina/executors/__init__.py [hunk 3]
```diff
@@ -216,6 +217,25 @@ def post_init(self):
             All class members created here will NOT be serialized when calling :func:`save`. Therefore if you
             want to store them, please override the :func:`__getstate__`.
         """
+        self._set_device()
```
[the remark this answers is not in the document]
Comment (nan-wang) [reply jina/executors/__init__.py:220]: fixed

File: jina/executors/__init__.py [hunk 4]
```diff
@@ -536,3 +556,82 @@ def __call__(self, req_type, *args, **kwargs):
...
+        super().post_init()
+        self.pre_set_device()
+        self.build_model()
+        self.post_set_device()
+
+    def build_model(self):
+        """
+        Build the model with the framework set by `self._backend`.
+        """
+        raise NotImplementedError
```
Comment (hanxiao) [inline jina/executors/__init__.py:579]:
```suggestion

```

File: jina/executors/__init__.py (hunk 1 above)
Comment (hanxiao) [inline jina/executors/__init__.py:571]:
```suggestion

```

File: jina/executors/__init__.py [hunk 5]
```diff
@@ -536,3 +556,82 @@ def __call__(self, req_type, *args, **kwargs):
...
+        `tensorflow`, `pytorch`, `onnx`, and, `paddlepaddle`.
+
+    ..notes:
+        The derived classes must implement `build_model()` and `set_device()` methods.
+    """
+    def post_init(self):
+        super().post_init()
+        self.pre_set_device()
+        self.build_model()
+        self.post_set_device()
```
Comment (hanxiao) [inline jina/executors/__init__.py:573]: `pre_/post_set_device` should be replaced by one function with if branching and put it into `_post_init_wrapper`

File: jina/executors/__init__.py [hunk 6]
```diff
@@ -536,3 +556,82 @@ def __call__(self, req_type, *args, **kwargs):
...
+
+    def post_set_device(self):
+        """
+        Set the device on which the model will be executed after building the model.
+
+        ..notes:
+            In the case of using GPUs, we only use the first gpu from the visible gpus. To specify which gpu to use,
+            please use the environment variable `CUDA_VISIBLE_DEVICES`.
+        """
+        pass
```
Comment (hanxiao) [inline jina/executors/__init__.py:599]:
```suggestion
    def set_device(self):
        """
        Set the device on which the model will be executed before building the model.

        ..notes:
            In the case of using GPUs, we only use the first gpu from the visible gpus. To specify which gpu to use,
            please use the environment variable `CUDA_VISIBLE_DEVICES`.
        """
        pass

```

File: jina/executors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nan-wang) [reply jina/executors/__init__.py:220]: fixed

File: jina/executors/encoders/image/onnx.py [hunk 7]
```diff
@@ -42,7 +44,7 @@ def __init__(self,
         self.raw_model_path = model_path
         self.model_name = ""
 
-    def post_init(self):
+    def build_model(self):
```
Comment (hanxiao) [inline jina/executors/encoders/image/onnx.py:47]:
```suggestion
    def post_init(self):
```

Comment (nan-wang) [reply jina/executors/encoders/image/onnx.py]: fixed

File: jina/executors/__init__.py (hunk 5 above)
Comment (nan-wang) [reply jina/executors/__init__.py]: This is a great idea saving lots of codes!

File: jina/executors/__init__.py [hunk 8]
```diff
@@ -536,3 +547,221 @@ def __call__(self, req_type, *args, **kwargs):
...
+
+        :param output_feature: the name of the layer for feature extraction.
+        :param model_path: the path of the model in the format of `.onnx`. Check a list of available pretrained
+            models at https://github.com/onnx/models#image_classification and download the git LFS to your local path.
+            The ``model_path`` is the local path of the ``.onnx`` file, e.g. ``/tmp/onnx/mobilenetv2-1.0.onnx``.
+        """
+        super().__init__(*args, **kwargs)
+        self.outputs_name = output_feature
+        self.raw_model_path = model_path
+        self.model_name = ""
```
Comment (hanxiao) [inline jina/executors/__init__.py:736]:
```suggestion
        self.model_name = ''
```

File: jina/executors/encoders/nlp/flair.py [hunk 9]
```diff
@@ -76,4 +76,11 @@ def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
         from flair.embeddings import Sentence
         c_batch = [Sentence(row) for row in data]
         self.model.embed(c_batch)
-        return torch.stack([c_text.get_embedding() for c_text in c_batch]).detach().numpy()
+        result = torch.stack([c_text.get_embedding() for c_text in c_batch]).detach()
+        if self.on_gpu:
+            result = result.cpu()
+        return result.numpy()
```
Comment (hanxiao) [inline jina/executors/encoders/nlp/flair.py:82]: this part can still be improved. but let's leave it to the next PR

**Pull request merged**